tapir/coop/services/ResignMemberService.py

```diff
@@ -0,0 +1,84 @@
+import datetime
+
+from tapir.accounts.models import TapirUser
+from tapir.coop.models import ResignedMembership, ShareOwnership
+from tapir.coop.forms import MembershipCancelForm
+from tapir.shifts.models import (
+    ShiftAttendanceTemplate,
+    ShiftAttendance,
+    ShiftAccountEntry,
+)
+from tapir.shifts.services.frozen_status_service import FrozenStatusService
+from tapir.shifts.models import ShiftUserData, ShiftAttendanceMode
+
+from tapir.utils.shortcuts import get_timezone_aware_datetime
+
+
+class ResignMemberService:
+    @staticmethod
+    def update_shifts_and_shares(
+        form: MembershipCancelForm, member: ResignedMembership
+    ):
+        tapir_user: TapirUser = getattr(member.share_owner, "user", None)
+        if not tapir_user:
+            print("Couldn't find an existing Tapir user.")
+            return
+        shares = ShareOwnership.objects.filter(share_owner=member.share_owner)
+        if (
+            member.willing_to_gift_shares_to_coop
+            or member.transfering_shares_to is not None
+        ):
+            for share in shares:
+                if share.is_fully_paid:
+                    if member.transfering_shares_to is not None:
+                        ShareOwnership.objects.create(
+                            share_owner=member.transfering_shares_to,
+                            amount_paid=share.amount_paid,
+                            start_date=member.cancellation_date,
+                        )
+                        share.delete()
+                    else:
+                        share.is_active = False
+                else:
+                    raise ValueError(
+                        "Shareowner didn't fully paid all shares, so it cannot be tranfered to someone else."
+                    )
+                    break
+
+            for attendance_template in ShiftAttendanceTemplate.objects.filter(
+                user=tapir_user
+            ):
+                start_date = get_timezone_aware_datetime(
+                    member.cancellation_date, datetime.time()
+                )
+                attendance_template.cancel_attendances(start_date)
+                attendance_template.delete()
+
+                attendances = ShiftAttendance.objects.filter(
+                    user=tapir_user,
+                    slot__shift__start_time__gte=start_date,
+                    state=ShiftAttendance.State.PENDING,
+                )
+
+                for attendance in attendances:
+                    attendance.state = ShiftAttendance.State.CANCELLED
+                    attendance.save()
+
+                for shift_account_entry in ShiftAccountEntry.objects.filter(
+                    user=tapir_user
+                ):
+                    shift_account_entry.delete()
+
+                for shift_user_data in ShiftUserData.objects.filter(user=tapir_user):
+                    shift_user_data.attendance_mode = (ShiftAttendanceMode.FROZEN,)
```

Comment:
1. The is only one ShiftUserData per user because `ShiftUserData.user` is a `OneToOneField`, so you can do ``shift_user_data = ShiftUserData.objects.get(...)` instead of looping over the filter results
2. Here you only change the variable but don't `save()` it back to the database
3. tapir/shifts/management/commands/run_freeze_checks.py is run one every night. So if the member is not below -4 points, they will get unfrozen anyway. Also, freezing them would affect the statistics about how many members are frozen, so I think we should leave the attendance mode untouched.